### PR TITLE
Detect if access to localStorage is forbidden by the user's browser

### LINKED ISSUE #55079: Can't expand/contract Rustdoc's [-] symbol if "cookies" disabled
The output from `rustdoc`, at the top of the page for a `struct` or a `const`, etc., includes a line like,

> Struct crate::MyStruct    [-][src]

Clicking the `[-]` on a browser w/ restrictive cookie settings (which typically includes `localStorage`) results in an error similar to the following:

> `storage.js:11` Uncaught DOMException: Failed to read the '`localStorage`' property from '`Window`': Access is denied for this document.
>    at updateLocalStorage (file:///home/[...]/target/doc/storage.js:11:343)
>    at HTMLAnchorElement.toggleAllDocs (file:///home/[...]/target/doc/main.js:13:997)

Additionally, **the section fails to expand or contract**, as the exception isn't handled.

`rustdoc` is attempting to store the collapse/expand state, s.t. later visits to the page presumably maintain that state for the user's convenience.

I browse the web with an admittedly somewhat atypical setup: I whitelist cookies. `localStorage`, in most browsers, counts, since it is a similar mechanism of persistence. I think Rust's use of `localStorage` here is fine, but I would request that, if `localStorage` is forbidden by the user agent, that the documentation gracefully "degrade" to just expanding or contracting the requested item, but not persist it. As it is, the JS crashes while storing the state, and appears to never make it to actually expanding/contracting the item.

### PR DESCRIPTION
**Pull request by thanatos**
If the user's cookie/persistent storage setting forbid access to `localStorage`, catch the exception and abort the access.

Currently, attempting to use the expand/contract links at the top of the page for structs/consts/etc. fails due to an unhandled error while accessing `localStorage`, if such access is forbidden, as the exception from the failed access propagates all the way out, interrupting the expand/contract. Instead, I would like to degrade gracefully; the access won't happen (the collapse/expand state won't get persisted) but the actual expanding/contracting of the item will go on to succeed.

Fixes #55079

### CODE REVIEW

**Comment by thanatos** [comment]
Note: I'm not sure how to actually test this change, so I haven't done that yet.

**Comment by GuillaumeGomez** [comment]
Thanks for the PR! Please update the "getCurrentValue" function (just below the one you updated) too.

**Comment by thanatos** [comment]
Given the way `getCurrentValue` was written, it was much easier to update that function if the check itself was a separate function, so I've extracted the code to a separate function. (Which, honestly, I think was the right thing overall, and makes the update function nicer too.)

LMK if there's anything else needed; also, feel free to squash/not squash however the rust-lang folks like to do.

I ran `cargo doc` in a project of mine and replaced most of the `storage.js` that `cargo` emitted w/ the file as it is in my revision. (The exception being that the one `cargo` emits contains,
```javascript
var resourcesSuffix="";
```
just prior to the copyright banner; I have no idea where that comes from, so I left it for my test, only replacing the stuff below the copyright banner. If there's some more rigorous test I could perform, LMK, and I'll do that too.)

File: src/librustdoc/html/static/storage.js [hunk 1]
```diff
@@ -27,20 +27,38 @@ function onEach(arr, func) {
...
 
 function getCurrentValue(name) {
-    if (typeof(Storage) !== "undefined" && localStorage[name] !== undefined) {
+    if (usableLocalStorage() && localStorage[name] !== undefined) {
         return localStorage[name];
     }
     return null;
 }
 
+function usableLocalStorage() {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/storage.js]: Can move this function above the two others please? It's just me being careful but better be safe than sorry. :)

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/storage.js]: Try this code then:

```
foo();

function foo() {
    console.log('hello');
}
```

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+ rollup

**Comment by kornelski** [comment]
I think it'd be safer and more future-proof to wrap `getCurrentValue` and `updateLocalStorage` body `try`/`catch`.

For example, if browsers implemented read-only localStorage current check would pass, but write could still throw.

**Comment by GuillaumeGomez** [comment]
The write can only be performed on the setting page, therefore it's not really an issue if it doesn't work. The setting will just not get updated.

**Pull request merged**